src/rnn.jl

```diff
@@ -675,52 +738,84 @@ function rnntest(r::RNN, ws, x, hx=nothing, cx=nothing;
+                ot = sigm.(Woxt .+ Ro' * hs[i] .+ bWo .+ bRo)
+                cn = tanh.(Wcxt .+ Rc' * hs[i] .+ bWc .+ bRc)
+                cs[i] = ft .* cs[i] .+ it .* cn
+                hs[i] = ot .* tanh.(cs[i])
             end
-            push!(ys, hs[L])
         end
     elseif r.mode == 3           # GRU
         @assert r.inputMode == 0 || all(w[1:3*(1+r.direction)] .== nothing)
         # rt = σ(Wrxt + Rrht-1 + bWr + bRr)
         # it = σ(Wixt + Riht-1 + bWi + bRu)
         # h't = tanh(Whxt + rt◦(Rhht-1 + bRh) + bWh)
         # ht = (1 - it)◦h't + it◦ht-1
-        for t = 1:T
-            for l = 1:L
-                Wr,Wi,Wh,Rr,Ri,Rh = w[1+6*(l-1):6l]
-                bWr,bWi,bWh,bRr,bRi,bRh = w[6L+1+6*(l-1):6L+6l]
-                Wrxt = (l > 1 ? Wr' * hs[l-1] : r.inputMode==0 ? Wr' * x[:,:,t] : x[:,:,t])
-                Wixt = (l > 1 ? Wi' * hs[l-1] : r.inputMode==0 ? Wi' * x[:,:,t] : x[:,:,t])
-                Whxt = (l > 1 ? Wh' * hs[l-1] : r.inputMode==0 ? Wh' * x[:,:,t] : x[:,:,t])
-                rt = sigm.(Wrxt .+ Rr' * hs[l] .+ bWr .+ bRr)
-                it = sigm.(Wixt .+ Ri' * hs[l] .+ bWi .+ bRi)
-                ht = tanh.(Whxt .+ rt .* (Rh' * hs[l] .+ bRh) .+ bWh)
-                hs[l] = (1 .- it) .* ht .+ it .* hs[l]
+        if direction == 0
+            for t = 1:T
+                for l = 1:L
+                    Wr,Wi,Wh,Rr,Ri,Rh = w[1+6*(l-1):6l]
+                    bWr,bWi,bWh,bRr,bRi,bRh = w[6L+1+6*(l-1):6L+6l]
+                    Wrxt = (l > 1 ? Wr' * hs[l-1] : r.inputMode==0 ? Wr' * x[:,:,t] : x[:,:,t])
+                    Wixt = (l > 1 ? Wi' * hs[l-1] : r.inputMode==0 ? Wi' * x[:,:,t] : x[:,:,t])
+                    Whxt = (l > 1 ? Wh' * hs[l-1] : r.inputMode==0 ? Wh' * x[:,:,t] : x[:,:,t])
+                    rt = sigm.(Wrxt .+ Rr' * hs[l] .+ bWr .+ bRr)
+                    it = sigm.(Wixt .+ Ri' * hs[l] .+ bWi .+ bRi)
+                    ht = tanh.(Whxt .+ rt .* (Rh' * hs[l] .+ bRh) .+ bWh)
+                    hs[l] = (1 .- it) .* ht .+ it .* hs[l]
+                end
+                push!(ys, hs[L])
+            end
+        else
+            bidirect() do X, i, ti, skip
+                Wr,Wi,Wh,Rr,Ri,Rh = w[1+6*(i-1):6i]
+                bWr,bWi,bWh,bRr,bRi,bRh = w[6L+1+6*(i-1):6L+6i]
+                Wrxt = skip ? X[:, :, ti] : Wr' * X[:, :, ti]
+                Wixt = skip ? X[:, :, ti] : Wi' * X[:, :, ti]
+                Whxt = skip ? X[:, :, ti] : Wh' * X[:, :, ti]
+                rt = sigm.(Wrxt .+ Rr' * hs[i] .+ bWr .+ bRr)
+                it = sigm.(Wixt .+ Ri' * hs[i] .+ bWi .+ bRi)
+                ht = tanh.(Whxt .+ rt .* (Rh' * hs[i] .+ bRh) .+ bWh)
+                hs[i] = (1 .- it) .* ht .+ it .* hs[i]
             end
-            push!(ys, hs[L])
         end
     else
         error("RNN not supported")
     end
-    y = reshape(hcat(ys...), ysize)
+    y = r.direction == 0 ? reshape(hcat(ys...), ysize) : ys
```

Comment:
You don't put the output in the right order?  We want our output to match cudnn exactly...